cluster/core/src/main/scala/com/lightbend/lagom/internal/cluster/JoinClusterImpl.scala

```diff
@@ -23,6 +23,11 @@ private[lagom] object JoinClusterImpl {
     // join self if seed-nodes are not configured in dev-mode,
     // otherwise it will join the seed-nodes automatically
     val cluster = Cluster(system)
+
+    if (joinSelf) {
+      system.log.warning("Self-joining can potentially conflict with cluster formation and discovery.")
+    }
```

Comment:
Well, that is a much bigger and more complex task.

I think logging is fine as a quick improvement, but as described in #1337, I think it should only log in production mode.

@erip you can check the mode by injecting `Environment` into the constructor and looking at `environment.mode`.

For example: https://github.com/lagom/lagom/blob/adb37ef1127d20964cdf97d846206e263e43a0d3/service/scaladsl/api/src/main/scala/com/lightbend/lagom/scaladsl/api/deser/ExceptionSerializer.scala#L57-L65

I think it would also be a good idea to adjust the message to be more specific about what to change and why. In the case that originally triggered this issue, the users didn't even realize that they had configured this setting, or know what "self-joining" is! It was just accidentally copied and pasted from another example. Including the specific property name would help give users in this situation something they can search for in their code.

Maybe something like:

> The "lagom.cluster.join-self" setting should not be enabled in production, because it can conflict with Akka Cluster Bootstrap or cause split-brain clusters.

If you want to get fancy, you can also lookup the [`ConfigOrigin`](https://github.com/lightbend/config/blob/f6680a5dad51d992139d45a84fad734f1778bf50/config/src/main/java/com/typesafe/config/ConfigOrigin.java#L11-L15) for the setting, to print out where it was set in the first place!

Reply:
Right, we try to keep implementation classes private just for that reason! 😄

Reply:
This is kind of weird. By adding a `play.api.Environment` param to `JoinClusterImpl#join`, this forces an injection of `play.api.Environment` into `ClusterComponents`, but it's traits all the way down. (`trait ReadSideXyzPersistenceComponents <: trait ReadSidePersistenceComponents <: trait WriteSidePersistenceComponents <: trait ClusterComponents`). How do I inject into the components? I don't think it's the responsibility of the user to mix-in the `play.api.Environment`, so that shouldn't appear in the public API.

Reply:
@erip you should declare an abstract `def environment: Environment` to `ClusterComponents`, just like the `actorSystem` abstract method that's already there.

`play.api.Environment` is ultimately contributed by Play's [`BuiltInComponentsFromContext`](https://github.com/playframework/playframework/blob/79cc59b6cce3840a9f40061565426b28bb79c681/framework/src/play/src/main/scala/play/api/ApplicationLoader.scala#L122-L123).

Reply:
There was the missing piece. :) And I suppose when testing, the `Environment` can either be mocked or stubbed out.